### PR TITLE
[MRG] Just in time SAGA.

### PR DESCRIPTION
**Pull request by zermelozf**
A squashed version of #38 ontaining:
- SAGA algorithm in cython.
- Basic python version of SAG and SAGA for testing.
- Support for proximity operators through the `Penalty` base class.
- L1 proximity operator with just in time update for sparse data.

### CODE REVIEW

**Comment by fabianp** [comment]
Please add a brief  description of the contribution so that people seeing this for the first time know what is implemented (ie saga algorithm + support for prox + jit updates for sparse data + test... )

File: lightning/impl/sag_fast.pyx [hunk 1]
```diff
@@ -12,13 +13,86 @@ cimport numpy as np
...
+        for jj in xrange(n_nz):
+            j = indices[jj]
+            w[j] = fmax(w[j] - stepsize * w_scale, 0) \
+                    - fmax(-w[j] - stepsize * w_scale, 0)
+
+    cdef void projection_lagged(self,
+                                double* w,
+                                int* indices,
+                                double stepsize,
+                                double* w_scale,
```
Comment (mblondel) [inline lightning/impl/sag_fast.pyx]: w_scale isn't used. Does it mean that you don't support elastic net?

Comment (fabianp) [inline lightning/impl/sag_fast.pyx]: It does, the scale is used through scale_cum. The w_scale is maintained in case some prox wants to overwrite it.

Comment (mblondel) [inline lightning/impl/sag_fast.pyx]: Thanks for the clarification. I would remove w_scale from the function signature until we have an actual use case. Also, I don't see any test for elastic net so this means that scale_cum is not tested.

File: lightning/impl/sag_fast.pyx [hunk 2]
```diff
@@ -12,13 +13,86 @@ cimport numpy as np
...
+            j = indices[jj]
+            incr_scale = scale_cumm[t] - scale_cumm[last[j]]
+            w[j] = fmax(w[j] - stepsize * incr_scale, 0) \
+                    - fmax(-w[j] - stepsize * incr_scale, 0)
+            last[j] = t
+
+
+    cdef double regularization(self, np.ndarray[double, ndim=1] coef):
+
+        cdef int j, jj
```
Comment (mblondel) [inline lightning/impl/sag_fast.pyx]: `jj` is not used.

**Comment by fabianp** [comment]
This might also be of interest to @adefazio @agramfort @TomDLT

**Comment by agramfort** [comment]
thx

how does it compare in terms of perf with sklearn SAG version?

**Comment by zermelozf** [comment]
@agramfort We have updated [this gist](https://gist.github.com/zermelozf/d4670d6cffea09f6e6f3) with sklearn's `LogisticRegression`.

|   Multiclass (OVR) | time (sec.) | score |
|------------------------- |--------|---------|
|SAGClassifier         |8.727|0.1676|
|SAGAClassifier       |12.245|0.1757|
|LogisticRegression |12.061|0.1750|

I have assumed that lightning does OVR for multi class problems (@mblondel correct me if I misunderstood).

The score discrepancy seems to be caused by the different stopping criteria. We have not (yet) rigorously compared the convergence speed (to the optimum) of SAG vs SAGA. The advantage of SAGA over SAG is the possibility to specify an arbitrary proximity operator.

In term of code speed, adding `tol=0` leads to a runtime advantage for sklearn for 100 iterations over the whole 20newsgroup dataset.

|   100 iterations | time (sec.) | score |
|------------------------- |--------|---------|
|SAGClassifier         |198|0.174872665535|
|SAGAClassifier       |215|0.174872665535|
|LogisticRegression |117|0.174872665535|

**Comment by agramfort** [comment]
ok for the benefit of SAGA.

in terms of computation time what I read is that it's pretty much the same.

thx

**Comment by zermelozf** [comment]
@agramfort Yes, that is if we don't use any prox in SAGA as benchmarked above. Otherwise the jit update associated with the prox will slow the computation down.

**Comment by fabianp** [comment]
The interest of SAGA (for me) is in the support for composite loss functions. For smooth problems they read more or less the same in my experience.

**Comment by fabianp** [comment]
Also, comparing them is tricky since it ends up depending on how you choose the step size. Once we have adaptive step size (as described in the Schmidt paper) we could do more meaningful benchmarks

**Comment by mblondel** [comment]
Also IIRC sklearn's implementation of SAG uses a line search.

**Comment by TomDLT** [comment]
No, the sklearn's implementation of SAG uses a constant step size computed using the maximum Lipschitz constant over all samples (cf. [get_auto_step_size](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/sag.py#L17)).

**Comment by adefazio** [comment]
One trick that might help you speed SAGA up is to remove the random sampling of data points. Instead, at the beginning of the algorithm make a copy of the data set with the rows permuted at random (i.e. shuffled). Then every odd epoch access the datapoints in the original dataset in order, and every even epoch access them from the shuffled copy of the dataset in order also. This greatly reduces the number of TLB misses at the expense of requiring twice as much memory.
This trick only works for SAGA or SDCA, if you use it with SAG it will diverge.

**Comment by fabianp** [comment]
Thanks for the information @adefazio, I did not know about it. 

@zermelozf can your take into account Mathieu's comments?

**Comment by mblondel** [comment]
Interesting trick but intuitively it should work only for dense data. There is also the cost of copying the data which could take a few dozens of seconds for very large data. So not copying the data actually gives a head start.

**Comment by zermelozf** [comment]
I just removed `jj` in the last squashed commit.

**Comment by fabianp** [comment]
Can you remove the w_scale from projection_lagged as Mathieu suggested?

**Comment by fabianp** [comment]
Also, you need to add a test for elastic-net (where alpha != 0 AND beta != 0)

**Comment by zermelozf** [comment]
Elastic net test added and `w_scale` dependency removed in lagged update.

**Comment by mblondel** [comment]
Sounds great. Thanks for the awesome contrib. Are we good to go?

By the way, do you plan to add group lasso later?

**Comment by fabianp** [comment]
Good for me. Just needs to add SAGAClassifier and SAGARegression into `doc/references.rst`.

I do have a prox for group lasso penalty that I'm using. Will contribute once this is merged if you think its worth (I do think it would be a nice addition).

**Comment by mblondel** [comment]
+1

**Comment by zermelozf** [comment]
@fabianp @mblondel Doc reference updated.

**Comment by mblondel** [comment]
Alright, pressing the green button :)

**Comment by mblondel** [comment]
If you've got time, a SAGA vs. SDCA vs. Adagrad comparison for elastic net would be nice. You can build an example based on this:
http://www.mblondel.org/lightning/auto_examples/plot_l2_solvers.html

FYI SDCA doesn't work well without l2 regul (e.g. l1 regul only).

**Comment by zermelozf** [comment]
Yes, I was planning to do that as suggested by @fabianp as well. Thanks for the link, it seems like you have done 99% of the work already. I'll have a look at it next week after I finish a couple of things.

**Comment by fabianp** [comment]
Great work @zermelozf
